frontend/src/editor/editor.tsx

```diff
@@ -18,6 +18,135 @@ import { OnChangePlugin } from '@lexical/react/LexicalOnChangePlugin';
 
 import classes from './editor.module.css';
 
+
+function $getDocContext(): DocContext {
+  // Initialize default empty context
+  const docContext: DocContext = {
+    beforeCursor: '',
+    selectedText: '',
+    afterCursor: ''
+  };
+
+  // Get current selection
+  const selection = $getSelection();
+
+  // If no valid range selection exists, return empty context
+  if (!$isRangeSelection(selection)) {
+    return docContext;
+  }
+
+  // Get selected text content
+  docContext.selectedText = selection.getTextContent();
+
+  // Get points for traversal
+  let anchor = selection.anchor;
+  let focus = selection.focus;
+
+	// If the selection is backward, we need to swap the anchor and focus points.
+	if (selection.isBackward()) {
+		const temp = anchor;
+		anchor = focus;
+		focus = temp;
+	}
+
+  const anchorNode = anchor.getNode();
+  const focusNode = focus.getNode();
+  const anchorOffset = anchor.offset;
+  const focusOffset = focus.offset;
+
+  // Collect text before cursor
+  docContext.beforeCursor = getCursorText(anchorNode, anchorOffset, 'before');
+
+  // Collect text after cursor
+  docContext.afterCursor = getCursorText(focusNode, focusOffset, 'after');
+
+  return docContext;
+}
+
+// DFS traversal to get document order
+function collectNodes(node: any, visitedNodes: Set<string>, allNodes: LexicalNode[]) {
+	const nodeKey = node.getKey();
+	if (visitedNodes.has(nodeKey)) return;
+	visitedNodes.add(nodeKey);
+
+	allNodes.push(node);
+
+	// Add children in document order
+	if ('getChildren' in node) {
```

Comment:
Ah, there's your type-assertion. I wonder if we can find docs of when this happens.